data/shaders/conditionals.frag

```diff
@@ -2,7 +2,10 @@ varying vec4 dummy;
 
 void main(void)
 {
-    float d = fract(gl_FragCoord.x * gl_FragCoord.y * 0.0001);
+    // needs to be declared highp since the multiplication can overflow in
+    // mediump, particularly if mediump is implemented as fp16
+    highp vec2 FragCoord = gl_FragCoord.xy;
+    float d = fract(FragCoord.x * FragCoord.y * 0.0001);
```

Comment:
highp isn't guaranteed to be supported on ES2, so this will fail to compile on some drivers.

Reply:
Oh boy. That is an obscure corner of the spec I almost would rather not know about. It isn't clear to me there is any way to implement a lot of this that is guaranteed correct in ES2, so much is left unspecified.. Even if you factored out the multiplication to keep everything in spec, ESSL allows reassociation of floating point multipication if I recall correctly?

Reply:
I assume we're talking about:

>The vertex language requires any uses of lowp, mediump and highp to compile and link without error.
The fragment language requires any uses of lowp and mediump to compile without error. Support for
highp is optional.

My recollection is glmark2 already handles compile and link error failures (I saw it on broadcom's closed driver) and throws a kind of not-supported result, which would be the right answer.  No sense benchmarking incorrect rendering.

Reply:
@anholt: But you don't *know* that it's broken. Just because mediump isn't enough on one GPU, doesn't mean it's not enough on another GPU. One example would be Tegra 2, which uses FP20 for mediump. It's not enough for highp, but it's much better than FP16, which is what broke here.

Plus different GPUs have different FP16 handling, and some (like Mali-400) has best-in-class FP16 handling but no FP32 support, while others have terrible FP16 precision (like no denorms), but FP32 handling.

Reply:
> Oh boy. That is an obscure corner of the spec I almost would rather not know about. It isn't clear to me there is any way to implement a lot of this that is guaranteed correct in ES2, so much is left unspecified..

Just make it do what it used to do if  the `GL_FRAGMENT_PRECISION_HIGH` variable isn't defined. That way, you fix this where it's possible, but without breaking anything.

Reply:
One possible implementation would be this:
```glsl
#ifdef GL_FRAGMENT_PRECISION_HIGH 
#define maxp highp
#else
#define maxp mediump
#endif
```

...and then use `maxp` instead of `highp`.

Reply:
If you want to be able to use mediump on HW that has a more-precise-than-required mediump, you should be using `glGetShaderPrecisionFormat()` to query the accuracy of the mediump.

Reply:
If you want to be guaranteed correct results, sure. But don't let perfect be the enemy of good, and let's not pointlessly break things entirely just because we're not always guaranteed correct results.

Reply:
In an effort to get this unstuck, here's what I think this should look like instead:

```diff
diff --git a/data/shaders/conditionals.frag b/data/shaders/conditionals.frag
index 3bd2507..8b53b40 100644
--- a/data/shaders/conditionals.frag
+++ b/data/shaders/conditionals.frag
@@ -2,7 +2,12 @@ varying vec4 dummy;
 
 void main(void)
 {
-    float d = fract(gl_FragCoord.x * gl_FragCoord.y * 0.0001);
+#ifdef GL_FRAGMENT_PRECISION_HIGH
+    highp vec2 FragCoord = gl_FragCoord.xy;
+#else
+    vec2 FragCoord = gl_FragCoord.xy;
+#endif
+    float d = fract(FragCoord.x * FragCoord.y * 0.0001);
 
 $MAIN$
 
diff --git a/data/shaders/function.frag b/data/shaders/function.frag
index 3e3c74f..f352ac4 100644
--- a/data/shaders/function.frag
+++ b/data/shaders/function.frag
@@ -8,7 +8,12 @@ $PROCESS$
 
 void main(void)
 {
-    float d = fract(gl_FragCoord.x * gl_FragCoord.y * 0.0001);
+#ifdef GL_FRAGMENT_PRECISION_HIGH
+    highp vec2 FragCoord = gl_FragCoord.xy;
+#else
+    vec2 FragCoord = gl_FragCoord.xy;
+#endif
+    float d = fract(FragCoord.x * FragCoord.y * 0.0001);
 
 $MAIN$
 
diff --git a/data/shaders/loop.frag b/data/shaders/loop.frag
index 31ae23e..0e7d95f 100644
--- a/data/shaders/loop.frag
+++ b/data/shaders/loop.frag
@@ -3,7 +3,12 @@ uniform int FragmentLoops;
 
 void main(void)
 {
-    float d = fract(gl_FragCoord.x * gl_FragCoord.y * 0.0001);
+#ifdef GL_FRAGMENT_PRECISION_HIGH
+    highp vec2 FragCoord = gl_FragCoord.xy;
+#else
+    vec2 FragCoord = gl_FragCoord.xy;
+#endif
+    float d = fract(FragCoord.x * FragCoord.y * 0.0001);
 
 $MAIN$
 
diff --git a/data/shaders/terrain-noise.frag b/data/shaders/terrain-noise.frag
index 7fea5c0..8011b65 100644
--- a/data/shaders/terrain-noise.frag
+++ b/data/shaders/terrain-noise.frag
@@ -17,7 +17,11 @@ uniform float time;
 uniform MEDIUMP vec2 uvScale;
 varying vec2 vUv;
 
+#ifdef GL_FRAGMENT_PRECISION_HIGH
+vec4 permute(highp vec4 x)
+#else
 vec4 permute(vec4 x)
+#endif
 {
     return mod(((x * 34.0) + 1.0) * x, 289.0);
 }
diff --git a/data/shaders/terrain.frag b/data/shaders/terrain.frag
index 84d085c..8c2e15f 100644
--- a/data/shaders/terrain.frag
+++ b/data/shaders/terrain.frag
@@ -67,7 +67,12 @@ void main() {
     vec3 pointSpecular = vec3( 0.0 );
     for ( int i = 0; i < MAX_POINT_LIGHTS; i ++ ) {
         vec4 lPosition = viewMatrix * vec4( pointLightPosition[ i ], 1.0 );
+
+#ifdef GL_FRAGMENT_PRECISION_HIGH
         vec3 lVector = lPosition.xyz + vViewPosition.xyz;
+#else
+        highp vec3 lVector = lPosition.xyz + vViewPosition.xyz;
+#endif
         float lDistance = 1.0;
         if ( pointLightDistance[ i ] > 0.0 )
             lDistance = 1.0 - min( ( length( lVector ) / pointLightDistance[ i ] ), 1.0 );
```

Reply:
Edited as @kusma suggested. If we want to refuse to run these tests on hardware with insufficiently precise mediump and no highp support, that can come as a follow on. I think this change is enough to fix the behaviour on some in-spec drivers without regressing anyone.